### PR TITLE
event check in parallel

### PR DESCRIPTION
**Pull request by leviok**
I have found an issue during sr2. tscache will stuck when number of deals increasing. 
when event checking takes too long, the tipsets from chainnotify will stuck in buffer. then tipset cache will behind the chain.
in that case, deal process will get wrong deal-state. such as can't find deal on chain when deal was published.
i use convenience and tricky way in this PR to handle this issue temporary.
i think we need to reconsider the adaptability of the TSC mechanism under high throughput.

### CODE REVIEW

**Review by magik6k — COMMENTED** [review]
I think this is safe to do, we just need to make sure that calling matchFn in thread-safe

File: chain/events/events_called.go [hunk 1]
```diff
@@ -376,17 +391,24 @@ func (we *watcherEvents) checkStateChanges(oldState, newState *types.TipSet) map
...
+				log.Errorf("event diff fn failed: %s", err)
+				wg.Done()
+				return
+			}
 
-		if ok {
-			res[tid] = data
-		}
+			if ok {
+				res[tid] = data
```
Comment (magik6k) [inline chain/events/events_called.go:406]: This map access needs a lock

**Review by magik6k — COMMENTED** [review]
Just 1 comment

File: chain/events/events_called.go (hunk 1 above)
Comment (whyrusleeping) [reply chain/events/events_called.go:406]: could use an array instead of a map and lock

**Comment by ZenGround0** [comment]
I spent some time learning about this solution and this part of the system in general and I think we should close this.

The primary reason is that @leviok points out that this is a temporary fix that should probably be addressed at the root by designing the tipset cache for high throughput (of access for deals?).  If the tipset cache is still getting stuck we should address the problem completely.

Even if we think going down the path of parallelizing the match function calls is the right thing to do in the short term this approach seems not to scale properly.  All provider match functions and all client match functions begin by calling a global lock.  If I understand right that there is one client/provider storageadapter per lotus node this at most halves the match function time and so we should either remove contention or dispatch match functions to one of only two go routines instead of spawning many contending goroutines.

**Pull request closed without merging**